api/server/router/network/network_routes.go

```diff
@@ -236,14 +245,133 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
 			return err
 		}
 		nw = &types.NetworkCreateResponse{
-			ID:      id,
-			Warning: warning,
+			ID: id,
 		}
 	}
 
+	nw.Warning = strings.Join(warnings, ". ") + "."
 	return httputils.WriteJSON(w, http.StatusCreated, nw)
 }
 
+type ipFamily int
+
+const (
+	ip4 ipFamily = 4
+	ip6 ipFamily = 6
+)
+
+func validateIPAM(ipam *network.IPAM, enableIPv6 bool) *multierror.Error {
```

Comment:
I'm a bit on the fence on having all this validation code in the API router. Given that we're already importing libnetwork, perhaps it'd make sense to move the validation code there (not sure where the right place would be inside libnetwork, perhaps in the `ipam`  or `ipamapi` package.

We can either have the function exported (and call it here; `ipam.Validate(...)`), or perhaps as part of the final commit (which turns it into an error), move it into `backend.CreateNetwork()`. That way calling that function would also validate before creating (regardless of where it's called from).

Reply:
> I'm a bit on the fence on having all this validation code in the API router.

I think the best would be to add IPAM validation call to `(*Daemon).CreateNetwork()`. However, the following method from swarmkit would prevent converting errors into warnings when the network is created from a non-leader manager node, as it returns either the network created or an error, but not both: 

https://github.com/moby/swarmkit/blob/ad0f3ae162fa30d0710105211d37996d5ec38b27/api/control.pb.go#L6678-L6685

Also, there's already some sort of network validation in the daemon pkg (ie. `d.containerStart() -> d.initializeNetworking() -> d.allocateNetwork() -> d.connectToNetwork() -> d.updateNetworkConfig() -> validateNetworkingConfig()`).

> Given that we're already importing libnetwork, perhaps it'd make sense to move the validation code there

Currently the API only uses constants and errors from libnetwork packages. If we call IPAM validation helper from the API due to my previous point, these helpers would need to either:

1. Rely on the API type `IPAMConfig`, and thus import it into libnetwork -- that'd be a bad separation of concern IMO ;
2. Convert API type `IPAMConfig` into libnetwork `IpamConf`. And then, errors would have to either reference fields from the latter struct and re-translated to match the former -- that'd make the validation layer more complex with no real gains. Or modeled after `IPAMConfig`, but then we're back at bad separation of concerns ;

Also, I think we should make sure as much as possible that libnetwork data structures never end up in an invalid state -- that's the crux of the original issue I'm solving here. Maintaining libnetwork is way harder if we can't trust this rule is enforced. This also discards option 2.

Reply:
> Currently the API only uses constants and errors from libnetwork packages. If we call IPAM validation helper from the API due to my previous point, these helpers would need to either:
>
> - Rely on the API type IPAMConfig, and thus import it into libnetwork -- that'd be a bad separation of concern IMO ;
> - Convert API type IPAMConfig into libnetwork IpamConf. And then, errors would have to either reference fields from the latter struct and re-translated to match the former -- that'd make the validation layer more complex with no real gains. Or modeled after IPAMConfig, but then we're back at bad separation of concerns ;
> - Also, I think we should make sure as much as possible that libnetwork data structures never end up in an invalid state -- that's the crux of the original issue I'm solving here. Maintaining libnetwork is way harder if we can't trust this rule is enforced. This also discards option 2.

- Looking at the validation we have in this PR, we only need `create.IPAM`, and `create.EnableIPv6`.
- `backend.CreateNetwork()` is implemented by `daemon.CreateNetwork()` / `daemon.createNetwork()`
- .. which already imports `"github.com/docker/docker/libnetwork/ipamapi"`
- .. and has a `getIpamConfig()` function (which can produce an error for invalid configuration)

And on the calling site (`network_routes.go`), we already have some checking for error-types returned by the backend (e.g. `libnetwork.NetworkNameError`) to turn errors into warnings.

So my thinking is to

- have the validation either in the backend (`daemon.createNetwork()` -> `getIpamConfig()`) itself (or in the `ipamapi` package, and called from there)
- make it return a sentinel error (for example, `libnetwork.InvalidIPAMConfig`)
- ^^ Ideally that error would implement `errdefs.InvalidParam`

For the current release, we catch that (`libnetwork.InvalidIPAMConfig`), and turn them into a warning, and for the next release, we remove that check, and return the error as-is (producing a 4XX status if it implements `errdefs.InvalidParam`)


Would that work?